pkg/controller/s3/bucket.go

```diff
@@ -0,0 +1,216 @@
+		For(&v1beta1.Bucket{}).
+		Complete(managed.NewReconciler(mgr,
+			resource.ManagedKind(v1beta1.BucketGroupVersionKind),
+			managed.WithExternalConnecter(&connector{kube: mgr.GetClient(), newClientFn: s3.NewClient}),
+			managed.WithReferenceResolver(managed.NewAPISimpleReferenceResolver(mgr.GetClient())),
+			managed.WithInitializers(managed.NewNameAsExternalName(mgr.GetClient()), &tagger{kube: mgr.GetClient()}),
+			managed.WithLogger(l.WithValues("controller", name)),
+			managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name)))))
+}
+
+type connector struct {
+	kube        client.Client
+	newClientFn func(ctx context.Context, credentials []byte, region string, auth awsclients.AuthMethod) (s3.BucketClient, error)
+}
+
+func (c *connector) Connect(ctx context.Context, mg resource.Managed) (managed.ExternalClient, error) {
+	cr, ok := mg.(*v1beta1.Bucket)
+	if !ok {
+		return nil, errors.New(errUnexpectedObject)
+	}
+
+	p := &awsv1alpha3.Provider{}
+	if err := c.kube.Get(ctx, types.NamespacedName{Name: cr.Spec.ProviderReference.Name}, p); err != nil {
+		return nil, errors.Wrap(err, errGetProvider)
+	}
+
+	if aws.BoolValue(p.Spec.UseServiceAccount) {
+		bucketClient, err := c.newClientFn(ctx, []byte{}, p.Spec.Region, awsclients.UsePodServiceAccount)
+		return &external{client: bucketClient, kube: c.kube}, errors.Wrap(err, errCreateBucketClient)
+	}
+
+	if p.GetCredentialsSecretReference() == nil {
+		return nil, errors.New(errGetProviderSecret)
+	}
+
+	s := &corev1.Secret{}
+	n := types.NamespacedName{Namespace: p.Spec.CredentialsSecretRef.Namespace, Name: p.Spec.CredentialsSecretRef.Name}
+	if err := c.kube.Get(ctx, n, s); err != nil {
+		return nil, errors.Wrap(err, errGetProviderSecret)
+	}
+
+	bucketClient, err := c.newClientFn(ctx, s.Data[p.Spec.CredentialsSecretRef.Key], p.Spec.Region, awsclients.UseProviderSecret)
+	return &external{client: bucketClient, kube: c.kube}, errors.Wrap(err, errCreateBucketClient)
+}
+
+type external struct {
+	kube   client.Client
+	client s3.BucketClient
+}
+
+func (e *external) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
+	cr, ok := mg.(*v1beta1.Bucket)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errUnexpectedObject)
+	}
+	if _, err := e.client.HeadBucketRequest(&awss3.HeadBucketInput{Bucket: aws.String(meta.GetExternalName(cr))}).Send(ctx); err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(resource.Ignore(s3.IsNotFound, err), errHead)
+	}
+	clients := bucketclients.MakeClients(cr, e.client)
+	for _, client := range clients {
+		updated, err := client.ExistsAndUpdated(ctx)
+		if err != nil {
+			return managed.ExternalObservation{ResourceExists: true, ResourceUpToDate: false}, err
+		}
+		if updated == bucketclients.NeedsUpdate {
+			println("Sub-resource needs update")
+			return managed.ExternalObservation{ResourceExists: true, ResourceUpToDate: false}, err
+		}
+		if updated == bucketclients.NeedsDeletion {
```

Comment:
This is because we don't want to delete the entire bucket, instead this is called if the user indicates they want to delete some specific subresource (e.g., if I want to delete my CORS configuration I can just unset that variable in the spec). The controller can handle this and send the appropriate request via the SDK

Reply:
Hmm, got it. Seems that's why my suggestion of reusing `ExternalObservation` instead of `ResourceStatus` wouldn't really apply directly because the former doesn't return a clue about whether it should be deleted or not.